python/pyspark/sql/dataframe.py

```diff
@@ -1809,18 +1810,27 @@ def repartition(  # type: ignore[misc]
 
         Repartition the data into 10 partitions.
 
-        >>> df.repartition(10).rdd.getNumPartitions()
-        10
+        >>> df.repartition(10).explain()
+        == Physical Plan ==
```

Comment:
Hmm, I don't think displaying the physical plan is very user-friendly; it can be daunting for new Spark users who might not understand what `RoundRobinPartitioning(10)` means. Personally I prefer the original examples.

Just curious, is it feasible to support `getNumPartitions()` in Spark Connect? This API is actually pretty useful.

Reply:
Can we do sth like:

```python
len(set(spark.range(10).repartition(2).select(sf.spark_partition_id()).collect()))
```

Reply:
@allisonwang-db I found there was a discussion on `getNumPartitions` [here](https://github.com/apache/spark/pull/16708), it seems the arguments still stand today.

@HyukjinKwon good idea, let me have a try

Reply:
I found that current changes can reflect the explainations like

```
Repartition the data into 7 partitions by 'age' and 'name columns.
```

```
        Repartition the data into 2 partitions by range in 'age' column.
        For example, the first partition can have ``(14, "Tom")``, and the second
        partition would have ``(16, "Bob")`` and ``(23, "Alice")``.
```

while if we use the numPartitions (not matter `rdd.getNumPartitions` or `spark_partition_id()`) in the examples, it can not provide such information.

Reply:
One way for such cases is to use `foreachPartition` in that case with printing each out. The problem would be the output wouldn't be caught in the Python test (as they are printed out from a separate thread from JVM IIRC).

So, maybe we could do sth like:

```python
df = spark.createDataFrame([(14, "Tom"), (23, "Alice"), (16, "Bob")], ["age", "name"])
df.repartition(2, "age").foreachPartition(lambda it: print(list(it)))  # doctest: +SKIP
[Row(age=14, name='Tom')]
[Row(age=23, name='Alice'), Row(age=16, name='Bob')]
```